src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -23,128 +23,300 @@
+            // "Severity" is basically a measure of compaction activity (CASSANDRA-3722).
+            if (USE_SEVERITY)
+                score += getSeverity(entry.getKey());
+            // lowest score (least amount of badness) wins.
+            newScores.put(entry.getKey(), score);
+        }
+
+        return newScores;
     }
 
     /**
-     * Update configuration from {@link DatabaseDescriptor} and estart the update-scheduler and reset-scheduler tasks
+     * Update configuration from {@link DatabaseDescriptor} and restart the various scheduler tasks
      * if the configured rates for these tasks have changed.
      */
-    public void applyConfigChanges()
+    public void applyConfigChanges(int newDynamicUpdateInternal, int newDynamicLatencyProbeInterval, double newDynamicBadnessThreshold)
     {
-        if (dynamicUpdateInterval != DatabaseDescriptor.getDynamicUpdateInterval())
+        if (dynamicUpdateInterval != newDynamicUpdateInternal)
         {
-            dynamicUpdateInterval = DatabaseDescriptor.getDynamicUpdateInterval();
+            dynamicUpdateInterval = newDynamicUpdateInternal;
             if (DatabaseDescriptor.isDaemonInitialized())
             {
-                updateSchedular.cancel(false);
-                updateSchedular = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(update, dynamicUpdateInterval, dynamicUpdateInterval, TimeUnit.MILLISECONDS);
+                if (updateScoresScheduler != null)
+                    updateScoresScheduler.cancel(false);
+                updateScoresScheduler = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(this::updateScores, dynamicUpdateInterval, dynamicUpdateInterval, TimeUnit.MILLISECONDS);
             }
         }
 
-        if (dynamicResetInterval != DatabaseDescriptor.getDynamicResetInterval())
+        if (this.dynamicLatencyProbeInterval != newDynamicLatencyProbeInterval)
         {
-            dynamicResetInterval = DatabaseDescriptor.getDynamicResetInterval();
+            this.dynamicLatencyProbeInterval = newDynamicLatencyProbeInterval;
             if (DatabaseDescriptor.isDaemonInitialized())
             {
-                resetSchedular.cancel(false);
-                resetSchedular = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(reset, dynamicResetInterval, dynamicResetInterval, TimeUnit.MILLISECONDS);
+                if (updateSamplesScheduler != null)
+                    updateSamplesScheduler.cancel(false);
+
+                if (newDynamicLatencyProbeInterval > 0)
+                {
+                    probeRateLimiter.setRate(dynamicLatencyProbeInterval);
+                    updateSamplesScheduler = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(this::updateSamples, dynamicLatencyProbeInterval, dynamicLatencyProbeInterval, TimeUnit.MILLISECONDS);
```

Comment:
I considered that but didn't like it because then we can't block in the acquire of the ratelimiter tokens (as that would block the logical clock increments). I think that the amount of work we're doing here is on the same order as before (iterating the samples map) so we should be ok right?

Reply:
Hmm... really I just wanted to not be blocked behind all the other things in that other thread pool. Where is an asynchronous rate limit when you need it. I suppose it's fine since we only need it to run on the order of seconds and we can stand to be delayed a little.